test/cuda/test_compilation_cache.cc

```diff
@@ -49,139 +78,238 @@ class OptionsCacheTest : public ::testing::Test {
+      lang::CanonicalTcString("kernel0"),
+      tc::makeTensorInfoVector(inputPtrs),
+      tc::makeTensorInfoVector(outputPtrs),
+      backendStr(),
+      options,
+      std::chrono::microseconds(10));
+  optionsCache->recordRuntime(
+      lang::CanonicalTcString("kernel0"),
+      tc::makeTensorInfoVector(inputPtrs),
+      tc::makeTensorInfoVector(outputPtrs),
+      backendStr(),
+      options,
+      std::chrono::microseconds(11));
+  optionsCache->recordRuntime(
+      lang::CanonicalTcString("kernel1"),
+      tc::makeTensorInfoVector(inputPtrs),
+      tc::makeTensorInfoVector(outputPtrs),
+      backendStr(),
+      options,
+      std::chrono::microseconds(1));
 
-  ret = tc::OptionsCache::getCache()->retrieveOptionsAndRuntimes(
-      "kernel1", inputPtrs, outputPtrs);
-  ASSERT_EQ(ret.size(), 1u);
-  ASSERT_EQ(ret[0].options, options);
-  ASSERT_EQ(ret[0].recordedRuntimes.size(), 1u);
-  ASSERT_EQ(ret[0].recordedRuntimes[0], std::chrono::microseconds(1));
+  {
+    OptionsCacheKey key{lang::CanonicalTcString("kernel0"),
+                        tc::makeTensorInfoVector(inputPtrs),
+                        tc::makeTensorInfoVector(outputPtrs),
+                        backendStr()};
+    ASSERT_EQ(optionsCache->count(key), 1u);
+    auto range = optionsCache->equal_range(key);
+    auto it = range.first;
+    ASSERT_EQ(it->second.mappingOptions, options);
+    ASSERT_EQ(it->second.runtimes.size(), 2u);
+    ASSERT_EQ(it->second.runtimes[0], std::chrono::microseconds(10));
+    ASSERT_EQ(it->second.runtimes[1], std::chrono::microseconds(11));
+  }
 
-  ret = tc::OptionsCache::getCache()->retrieveOptionsAndRuntimes(
-      "kernel2", inputPtrs, outputPtrs);
-  ASSERT_EQ(ret.size(), 0u);
+  {
+    OptionsCacheKey key{lang::CanonicalTcString("kernel1"),
+                        tc::makeTensorInfoVector(inputPtrs),
+                        tc::makeTensorInfoVector(outputPtrs),
+                        backendStr()};
+    auto range = optionsCache->equal_range(key);
+    auto it = range.first;
+    ASSERT_EQ(optionsCache->count(key), 1u);
+    ASSERT_EQ(it->second.mappingOptions, options);
+    ASSERT_EQ(it->second.runtimes.size(), 1u);
+    ASSERT_EQ(it->second.runtimes[0], std::chrono::microseconds(1));
+  }
 
-  ASSERT_EQ(tc::OptionsCache::getCache()->size(), 2u);
-  ASSERT_EQ(tc::OptionsCache::getCache()->totalSize(), 2u);
-  ASSERT_EQ(tc::OptionsCache::getCache()->numberAttemptedRetrievals, 3);
-  ASSERT_EQ(tc::OptionsCache::getCache()->numberSuccessfulRetrievals, 2);
-  ASSERT_EQ(tc::OptionsCache::getCache()->numberCacheAttemps, 3);
+  {
+    OptionsCacheKey key{lang::CanonicalTcString("kernel2"),
+                        tc::makeTensorInfoVector(inputPtrs),
+                        tc::makeTensorInfoVector(outputPtrs),
+                        backendStr()};
+    ASSERT_EQ(optionsCache->count(key), 0u);
+  }
+
+  ASSERT_EQ(optionsCache->size(), 2u);
+  ASSERT_EQ(optionsCache->numberAttemptedRetrievals, 0u);
```

Comment:
Are these counters still useful?  You have clearly attempted to retrieve something from the cache without them being incremented...  If there is another way to retrieve stuff from caches, we should not use the "unsafe" way here (and better remove it completely).  Otherwise, just drop those counters.